Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,339 @@
+derived downstream for `InfinitePlace` (which is a subtype of `AbsoluteValue`).
+
+## Main definitions
+ - `WithAbs` : type synonym equipping a semiring with an absolute value.
+ - `AbsoluteValue.completion` : the uniform space completion of a field `K` equipped with real
+  absolute value.
+ - `NumberField.InfinitePlace.completion` : the completion of a number field `K` at an infinite
+  place, obtained by completing `K` with respect to the absolute value associated to the infinite
+  place.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding` : the embedding `v.embedding : K →+* ℂ`
+  extended to `v.completion →+* ℂ`.
+ - `NumberField.InfinitePlace.Completion.extensionEmbedding_of_isReal` : if the infinite place `v`
+  is real, then this extends the embedding `v.embedding_of_isReal : K →+* ℝ` to
+  `v.completion →+* ℝ`.
+ - `NumberField.InfinitePlace.Completion.equiv_real_of_isReal` : the ring isomorphism
+  `v.completion ≃+* ℝ` when `v` is a real infinite place; the forward direction of this is
+  `extensionEmbedding_of_isReal`.
+ - `NumberField.InfinitePlace.Completion.equiv_complex_of_isComplex` : the ring isomorphism
+  `v.completion ≃+* ℂ` when `v` is a complex infinite place; the forward direction of this is
+  `extensionEmbedding`.
+
+## Main results
+ - `NumberField.Completion.locallyCompactSpace` : the completion of a number field at
+  an infinite place is locally compact.
+ - `NumberField.Completion.isometry_extensionEmbedding` : the embedding `v.completion →+* ℂ` is
+  an isometry. See also `isometry_extensionEmbedding_of_isReal` for the corresponding result on
+  `v.completion →+* ℝ` when `v` is real.
+ - `NumberField.Completion.bijective_extensionEmbedding_of_isComplex` : the embedding
+  `v.completion →+* ℂ` is bijective when `v` is complex. See also
+  `bijective_extensionEmebdding_of_isReal` for the corresponding result for `v.completion →+* ℝ`
+  when `v` is real.
+
+## Tags
+number field, embeddings, infinite places, completion
+-/
+noncomputable section
+
+/-- Type synonym equipping a semiring with an absolute value. -/
+@[nolint unusedArguments]
+def WithAbs {R S : Type*} [Semiring R] [OrderedSemiring S] :
+    AbsoluteValue R S → Type _ := fun _ => R
+
+namespace WithAbs
+
+variable {K : Type*} [Field K] (v : AbsoluteValue K ℝ)
+
+instance normedField : NormedField (WithAbs v) :=
+  v.normedField
+
+variable {L : Type*} [NormedField L] {f : WithAbs v →+* L} {v}
+
+/-- If the absolute value `v` factors through an embedding `f` into a normed field, then
+the distance associated to the absolute value also factors through `f`. -/
+theorem dist_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)
+    (x y : WithAbs v) :
+    dist x y = dist (f x) (f y) := by
+  rw [(normedField v).dist_eq, (inferInstanceAs <| NormedField L).dist_eq,
+    ← f.map_sub, h]
+  rfl
+
+instance : Inhabited (WithAbs v) := ⟨0⟩
```

Comment:
Do you need it here? It seems to me that you just use it below when speaking about completion.

Reply:
Not necessarily, but I think it should be in the `WithAbs` namespace. Have moved it to the top.